parquet-hadoop/src/test/java/org/apache/parquet/statistics/TestStatistics.java

```diff
@@ -0,0 +1,515 @@
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.math.BigInteger;
+import java.util.Iterator;
+import java.util.Random;
+
+import static junit.framework.Assert.assertEquals;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.*;
+import static org.apache.parquet.schema.Type.Repetition.OPTIONAL;
+
+public class TestStatistics {
+  private static final int MEGABYTE = 1 << 20;
+  private static final long RANDOM_SEED = System.currentTimeMillis();
+
+  public static class DataGenerationContext {
+    public static abstract class WriteContext {
+      protected final String path;
+      protected final Path fsPath;
+      protected final MessageType schema;
+      protected final int blockSize;
+      protected final int pageSize;
+      protected final boolean enableDictionary;
+      protected final boolean enableValidation;
+      protected final ParquetProperties.WriterVersion version;
+
+      public WriteContext(String path, MessageType schema, int blockSize, int pageSize, boolean enableDictionary, boolean enableValidation, ParquetProperties.WriterVersion version) throws IOException {
+        this.path = path;
+        this.fsPath = new Path(path);
+        this.schema = schema;
+        this.blockSize = blockSize;
+        this.pageSize = pageSize;
+        this.enableDictionary = enableDictionary;
+        this.enableValidation = enableValidation;
+        this.version = version;
+      }
+
+      public abstract void write(ParquetWriter<Group> writer) throws IOException;
+      public abstract void test() throws IOException;
+    }
+
+    public static void writeAndTest(WriteContext context) throws IOException {
+      File file = new File(context.path);
+
+      // Create the configuration, and then apply the schema to our configuration.
+      Configuration configuration = new Configuration();
+      GroupWriteSupport.setSchema(context.schema, configuration);
+      GroupWriteSupport groupWriteSupport = new GroupWriteSupport();
+
+      // Create the writer properties
+      final int blockSize = context.blockSize;
+      final int pageSize = context.pageSize;
+      final int dictionaryPageSize = pageSize;
+      final boolean enableDictionary = context.enableDictionary;
+      final boolean enableValidation = context.enableValidation;
+      ParquetProperties.WriterVersion writerVersion = context.version;
+      CompressionCodecName codec = CompressionCodecName.UNCOMPRESSED;
+
+      ParquetWriter<Group> writer = new ParquetWriter<Group>(context.fsPath, groupWriteSupport, codec, blockSize,
+        pageSize, dictionaryPageSize, enableDictionary, enableValidation, writerVersion, configuration);
+
+      context.write(writer);
+      writer.close();
+
+      context.test();
+      file.delete();
+    }
+  }
+
+  public static abstract class PagingValidator<T extends Comparable<T>> {
+    private static class ComputedStats<T extends Comparable<T>> {
```

Comment:
Sorry if I wasn't clear earlier: I don't think this should calculate stats at all.

Instead, the validation should get the min/max/numNulls and do this:
```java
T min = minFromStats();
T max = maxFromStats();
long numNullsFromStats = numNullsFromStats();
for (T value : readValue(reader)) {
  if (value != null) {
    assertTrue("min should be <= all values", compare(min, value) <= 0);
    assertTrue("min should be >= all values", compare(max, value) >= 0);
  } else {
    numNulls += 1;
  }
  assertEquals("number of nulls should match", numNullsFromStats, numNulls);
}
```